### PR TITLE
Forcing an embed means that it will always happen (Effects trapped posters, accidental consumption, and bullet shrapnel)

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Addresses #73105
I confidently asserted in the issue that the current behaviour was not a bug, but lo and behold when I went to investigate this on March 1st what did I find... a bug! Force embed actually never worked in the first place because the passed arguments were in the wrong order. Now they are named, so the order doesn't matter any more.
Also there was an argument called "silent" which did literally nothing, I just deleted it.

Sending the force flag will now mean that the forced embed _always_ embeds unless it is literally impossible. The only other place this effects is when you accidentally eat glass, so a similar case to this one.

While I was in there, I made posters _very slightly_ more discerning about what gloves block glass. Unfortunately there's no widely applied and useful trait like `THICK_MATERIAL` for gloves so I mostly just exluded latex and fingerless gloves... but that's better than nothing.

Anyway now if you bother to place a glass shard inside a poster and someone tries to pull it down and they're not wearing thin or fingerless gloves it will _always_ embed in their hand.

Oh and I put some balloon alerts in there.

**Unrelatedly to this change** fixing this seems to also fix an ancient bug where bullet shrapnel would roll its embed chance twice in a row, meaning that it had an (x/100)^2% chance of applying rather than an (x/100)% chance, significantly less than written. 
It is weird for that to go in this PR, but seems to be an effect of fixing this bug? Not sure there's a way of not fixing both at the same time.

## Why It's Good For The Game

You go to the effort of putting a glass shard behind a poster (the only way to associate your own fingerprints with the deed) and are lucky enough that the person who pulls it down isn't wearing gloves (almost everyone does) and it _still_ only has a 50% chance to embed the glass? 
The trap should work at least _some_ of the time.

## Changelog

:cl:
fix: Something which is "forced" to embed will now always actually embed. Resultingly, accidentally swallowing glass or encountering it behind a poster without adequate protection will now always cause it to embed into your body.
balance: Thin and fingerless gloves may no longer protect you from having glass in your hand after pulling down a poster
fix: Bullets containing shrapnel now use their correct chance to embed, which may cause bullet wounds to contain shrapnel more frequently.
/:cl:

### CODE REVIEW

File: code/datums/elements/embed.dm [hunk 1]
```diff
@@ -71,12 +71,16 @@
 	SIGNAL_HANDLER
 
 	if((!forced && blocked) || !istype(victim) || HAS_TRAIT(victim, TRAIT_PIERCEIMMUNE))
-		return
+		return FALSE
+
+	if(forced) // Skip all the other checks, we forced it
```
Comment (Mothblocks) [inline code/datums/elements/embed.dm:76]: This comment isn't really helpful or necessary

File: code/datums/elements/embed.dm [hunk 2]
```diff
@@ -71,12 +71,16 @@
 	SIGNAL_HANDLER
 
 	if((!forced && blocked) || !istype(victim) || HAS_TRAIT(victim, TRAIT_PIERCEIMMUNE))
```
Comment (Mothblocks) [inline code/datums/elements/embed.dm:73]: So forced isn't really forced? What is it?

File: code/game/objects/effects/poster.dm [hunk 3]
```diff
@@ -266,7 +265,7 @@
 
 	var/turf/user_drop_location = get_turf(user) //cache this so it just falls to the ground if they move. also no tk memes allowed.
 	if(!do_after(user, PLACE_SPEED, placed_poster, extra_checks = CALLBACK(placed_poster, TYPE_PROC_REF(/obj/structure/sign/poster, snowflake_wall_turf_check), src)))
-		to_chat(user, span_notice("The poster falls down!"))
+		balloon_alert(user, "interrupted!")
```
Comment (Mothblocks) [inline code/game/objects/effects/poster.dm:268]: Not necessary, since the do_after is already good enough visual feedback for this

**Comment by Jacquerel** [comment]
This failed CI task has revealed that my changes have reduced the chance of a bullet shot from a pistol embedding into you from 9%^2 to simply 9%. This is because it was rolling its embed chance and then, because `forced` did not do what it advertises, rolling it again a second time before confirming success. 
It seems to me that this means that test was never actually reliable because bullet embedding has always been RNG... but the failure had an extremely low chance to happen which I have now increased to almost 1 in 10.

I am unsure if this is a fix because it certainly _shouldn't_ have been rolling its embed twice but that doesn't change the fact that going from 0.8% to 9% chance to have a pistol bullet embed into you isn't a trivial increase.
And this presumably applies to many other projectile weapons.

Not totally sure what to do about that, I could preserve the original behaviour easily enough if we _want_ the embed chance to roll twice but that... seems incorrect.

Also either way I should do something about that unit test I guess?

**Comment by Jacquerel** [comment]
> This failed CI task has revealed that my changes have reduced the chance of a bullet shot from a pistol embedding into you from 9%^2 to simply 9%. This is because it was rolling its embed chance and then, because `forced` did not do what it advertises, rolling it again a second time before confirming success. It seems to me that this means that test was never actually reliable because bullet embedding has always been RNG... but the failure had an extremely low chance to happen which I have now increased to almost 1 in 10.

On further investigation this only seems to effect bullet _shrapnel_ which is not on every bullet, but is still on quite a lot of them.
Still, slightly less wide scope than I thought.

**Comment by Jacquerel** [comment]
Have pushed a commit which makes this behaviour work "as expected" and also fixes the unit test (I think), but that still means that what was meant to be a poster PR is coming with an unrelated balance change to "bullets in general", so let me know if you would like me to... do something else about that.

File: code/datums/elements/embed.dm [hunk 4]
```diff
@@ -172,28 +161,50 @@
...
+/datum/element/embed/proc/roll_embed_chance(obj/item/embedding_item, mob/living/victim, hit_zone, datum/thrownthing/throwingdatum)
+	var/actual_chance = embed_chance
+	var/penetrative_behaviour = 1 //Keep this above 1, as it is a multiplier for the pen_mod for determining actual embed chance.
+	if(embedding_item.weak_against_armour)
+		penetrative_behaviour = ARMOR_WEAKENED_MULTIPLIER
+
+	if(throwingdatum?.speed > embedding_item.throw_speed)
+		actual_chance += (throwingdatum.speed - embedding_item.throw_speed) * EMBED_CHANCE_SPEED_BONUS
+
+	if(!embedding_item.isEmbedHarmless()) // all the armor in the world won't save you from a kick me sign
```
Comment (san7890) [inline code/datums/elements/embed.dm:200]: can't we turn this into an early return

**Pull request merged**